### PR TITLE
Clean up ProtocolConformance::subst() and fix crash exposed by coroutines work

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes https://bugs.swift.org/browse/SR-8643 and simplifies some related code. Also addresses FIXMEs introduced by #18366.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@adrian-prantl Please take a look at the debug info change.

File: include/swift/SIL/TypeSubstCloner.h [hunk 1]
```diff
@@ -297,15 +297,12 @@ class TypeSubstCloner : public SILClonerWithScopes<ImplClass> {
...
-      // FIXME: It would be more elegant to run
-      // SubsMap.mapReplacementTypesOutOfContext() bup front, but it can assert.
-      Type Substitution = Type(ParamType).subst(SubsMap)->mapTypeOutOfContext();
-      return !Substitution->isOpenedExistential() &&
-             (Substitution->getCanonicalType() !=
-              ParamType->getCanonicalType());
-    });
+    for (auto ParamType : Sig->getSubstitutableParams()) {
+      if (!Type(ParamType).subst(SubsMap)->isEqual(ParamType))
+        return true;
```
Comment (adrian-prantl) [inline include/swift/SIL/TypeSubstCloner.h]: Nice!

**Review by DougGregor — COMMENTED** [review]
This is so much cleaner, thank you!

**Pull request merged**